apps/src/applab/applab.js

```diff
@@ -300,9 +304,40 @@ function queueOnTick() {
   window.setTimeout(Applab.onTick, getCurrentTickLength());
 }
 
+function updateDebugConsoleHistory(commandText) {
+  Applab.debugConsoleHistory.currentHistoryIndex = Applab.debugConsoleHistory.history.length + 1;
+  Applab.debugConsoleHistory.history[Applab.debugConsoleHistory.currentHistoryIndex - 1] = commandText;
+}
+
+function moveUpDebugConsoleHistory(currentInput) {
+  if (Applab.debugConsoleHistory.currentHistoryIndex > 0) {
+    Applab.debugConsoleHistory.currentHistoryIndex -= 1;
+  }
+  if (typeof Applab.debugConsoleHistory.history[Applab.debugConsoleHistory.currentHistoryIndex] !== 'undefined') {
+    return Applab.debugConsoleHistory.history[Applab.debugConsoleHistory.currentHistoryIndex];
+  }
+  return currentInput;
+}
+
+function moveDownDebugConsoleHistory(currentInput) {
+  if (Applab.debugConsoleHistory.currentHistoryIndex < Applab.debugConsoleHistory.history.length) {
+    Applab.debugConsoleHistory.currentHistoryIndex += 1;
+  }
+  if (Applab.debugConsoleHistory.currentHistoryIndex == Applab.debugConsoleHistory.history.length &&
+      currentInput == Applab.debugConsoleHistory.history[Applab.debugConsoleHistory.currentHistoryIndex - 1]) {
+    return '';
```

Comment:
Thanks for refactoring this! It looks great.  I've been trying the change out locally and noticed a funny edge case caused by this expression.

Suppose you have the following command history:

```
[0] continue
[1] step
[2] console.log(myVar)
```

Right after entering that last command your current history index `i` is 3.  Now consider two cases:

**Case one - editing most history entries**
* Press up twice, so that `i = 1` and the current input is "step"
* Edit the input, to something like "step in"
* Press down once.
* Now `i = 2` and the current input is "console.log(myVar)"  Your modified input is lost.

**Case two - editing the most recent history entry**
* Press up once so that `i = 2` and the current input is "console.log(myVar)"
* Edit the input, to something like "console.log(yourVar)"
* Press down once.
* Now `i = 3` and the current input is still "console.log(yourVar)" - unlike case one, we preserve your modified input.

I'm not sure which behavior is more desirable, to be honest, but I wanted to make sure you meant for this to be an edge case.  It's not totally clear to me what my zsh history is doing, but it starts out looking more like case two no matter which history entry I'm editing.  For students, I might be tempted to say we should always use case one so that the arrow keys "always work."  What do you think?